File: typescript/src/E2_3_4_5.ts
// yarn run build && node build/src/E2_3_4_5.js

const curry = <A, B, C>(f: (a: A, b: B) => C): ((a: A) => (b: B) => C) =>
  (a: A) => (b: B) => f(a, b);

const uncurry = <A, B, C>(f: (a: A) => (b: B) => C): ((a: A, b: B) => C) =>
  (a: A, b: B) => f(a)(b);

const compose = <A, B, C>(f: (b: B) => C, g: (a: A) => B): ((a: A) => C) =>
  (a: A) => f(g(a));

File: scala/E2_3_4_5.scala
// scala E2_3_4_5.scala

object E2_3_4_5 {
  def curry[A,B,C](f: (A, B) => C): A => (B => C) =
    a => b => f(a, b)

  def uncurry[A,B,C](f: A => B => C): (A, B) => C =
    (a, b) => f(a)(b)

  def compose[A,B,C](f: B => C, g: A => B): A => C =
    x => f(g(x)) 

  def main(args: Array[String]) = {}
}
